metaflow/runtime.py

```diff
@@ -219,14 +224,45 @@ def execute(self):
 
                 if time.time() - progress_tstamp > PROGRESS_INTERVAL:
                     progress_tstamp = time.time()
-                    msg = "%d tasks are running: %s." % (
-                        self._num_active_workers,
-                        "e.g. ...",
-                    )  # TODO
+                    tasks_print = ", ".join(
+                        [
+                            "%s (%d running; %d done)" % (k, v[0], v[1])
+                            for k, v in self._active_tasks.items()
+                            if k != 0 and v[0] > 0
+                        ]
+                    )
+                    if self._active_tasks[0] == 0:
+                        msg = "No tasks are running."
+                    else:
+                        if self._active_tasks[0] == 1:
+                            msg = "1 task is running: "
+                        else:
+                            msg = "%d tasks are running: " % self._active_tasks[0]
+                        msg += "%s." % tasks_print
+
                     self._logger(msg, system_msg=True)
-                    msg = "%d tasks are waiting in the queue." % len(self._run_queue)
+
+                    if len(self._run_queue) == 0:
+                        msg = "No tasks are waiting in the queue."
+                    else:
+                        if len(self._run_queue) == 1:
+                            msg = "1 task is waiting in the queue: "
+                        else:
+                            msg = "%d tasks are waiting in the queue." % len(
+                                self._run_queue
+                            )
+
                     self._logger(msg, system_msg=True)
-                    msg = "%d steps are pending: %s." % (0, "e.g. ...")  # TODO
+                    if len(self._unprocessed_steps) == 0:
+                        msg = "No steps are pending."
+                    else:
+                        if len(self._unprocessed_steps) == 1:
+                            msg = "1 step is pending: "
+                        else:
+                            msg = "%d steps are pending: " % len(
```

Comment:
What is the definition of a step being pending? Is a step pending if no task for that step has started? Or is it the case that not all tasks have been finished thus far? I am not sure of the utility of printing out pending steps in either of the cases though.

Reply:
Pending is steps that haven't started running. It gives an idea of where you are in the "execution frontier". I can rephrase to say %d steps not started